prover/src/constraints/evaluator.rs

```diff
@@ -23,18 +24,96 @@ use utils::{iterators::*, rayon};
 #[cfg(feature = "concurrent")]
 const MIN_CONCURRENT_DOMAIN_SIZE: usize = 8192;
 
-// CONSTRAINT EVALUATOR
+// CONSTRAINT EVALUATOR TRAIT
+// ================================================================================================
+
+pub trait ConstraintEvaluator<'a, E: FieldElement> {
+    type Air: Air<BaseField = E::BaseField>;
+
+    fn evaluate<T: TraceLde<BaseField = E::BaseField, ExtensionField = E>>(
+        self,
+        trace: &T,
+        domain: &'a StarkDomain<E::BaseField>,
+    ) -> ConstraintEvaluationTable<'a, E>;
+}
+
+// CONSTRAINT EVALUATOR DEFAULT IMPLEMENTATION
 // ================================================================================================
 
-pub struct ConstraintEvaluator<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> {
+pub struct DefaultConstraintEvaluator<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> {
     air: &'a A,
     boundary_constraints: BoundaryConstraints<E>,
     transition_constraints: TransitionConstraints<E>,
     aux_rand_elements: AuxTraceRandElements<E>,
     periodic_values: PeriodicValueTable<E::BaseField>,
 }
 
-impl<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> ConstraintEvaluator<'a, A, E> {
+impl<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> ConstraintEvaluator<'a, E>
+    for DefaultConstraintEvaluator<'a, A, E>
+{
+    type Air = A;
+
+    fn evaluate<T: TraceLde<BaseField = E::BaseField, ExtensionField = E>>(
```

Comment:
I think the implementation of this function is the same as the function which starts on line 153 below. We can probably just remove that function (and move the comment here).

Reply:
Good catch, thank you!